server/src/main/java/org/elasticsearch/snapshots/RestoreService.java

```diff
@@ -424,144 +410,212 @@ public ClusterState execute(ClusterState currentState) {
                         }
-                        indexMdBuilder.settings(Settings.builder()
-                                                    .put(snapshotIndexMetadata.getSettings())
-                                                    .put(IndexMetadata.SETTING_INDEX_UUID, currentIndexMetadata.getIndexUUID()));
-                        IndexMetadata updatedIndexMetadata = indexMdBuilder.index(renamedIndexName).build();
-                        rtBuilder.addAsRestore(updatedIndexMetadata, recoverySource);
-                        blocks.updateBlocks(updatedIndexMetadata);
-                        mdBuilder.put(updatedIndexMetadata, true);
-                        renamedIndex = updatedIndexMetadata.getIndex();
                     }
-                    for (int shard = 0; shard < snapshotIndexMetadata.getNumberOfShards(); shard++) {
-                        shardsBuilder.put(
-                            new ShardId(renamedIndex, shard),
-                            new RestoreInProgress.ShardRestoreStatus(clusterService.state().nodes().getLocalNodeId()));
+                    newIndexUUID = currentIndexMetadata.getIndexUUID();
+                    indexMdBuilder.settings(
+                        Settings.builder()
+                            .put(snapshotIndexMetadata.getSettings())
+                            .put(IndexMetadata.SETTING_INDEX_UUID, newIndexUUID));
+                    IndexMetadata updatedIndexMetadata = indexMdBuilder.index(renamedIndexName).build();
+                    rtBuilder.addAsRestore(updatedIndexMetadata, recoverySource);
+                    blocks.updateBlocks(updatedIndexMetadata);
+                    mdBuilder.put(updatedIndexMetadata, true);
+                    renamedIndex = updatedIndexMetadata.getIndex();
+                }
+                relationIndexUUIDs.compute(relationName, (_, indexUUIDs) -> {
+                    if (indexUUIDs == null) {
+                        indexUUIDs = new ArrayList<>();
+                    }
+                    indexUUIDs.add(newIndexUUID);
+                    return indexUUIDs;
+                });
+                for (int shard = 0; shard < snapshotIndexMetadata.getNumberOfShards(); shard++) {
+                    shardsBuilder.put(
+                        new ShardId(renamedIndex, shard),
+                        new RestoreInProgress.ShardRestoreStatus(clusterService.state().nodes().getLocalNodeId()));
+                }
+            }
+
+            for (var entry : relationIndexUUIDs.entrySet()) {
+                RelationName relationName = entry.getKey();
+                String newSchema = relationName.schema().replaceAll(
+                    request.schemaRenamePattern(),
+                    request.schemaRenameReplacement()
+                );
+                String newTableName = relationName.name().replaceAll(
+                    request.tableRenamePattern(),
+                    request.tableRenameReplacement()
+                );
```

Comment:
Didn't the rename already happen in this stage by the logic at `restoreSnapshot`?

Reply:
Kind of - the `indices` has entries from old name to renamed name, but I need the new `RelationName` here. Found it easier to just re-create it - otherwise I'd have to track it in the `relationIndexUUIDs somehow via an additional record/tuple or something.